discovery/db_discovery.go

```diff
@@ -0,0 +1,92 @@
+package discovery
+
+import (
+	"time"
+
+	"github.com/livepeer/go-livepeer/common"
+	"github.com/livepeer/go-livepeer/core"
+	lpTypes "github.com/livepeer/go-livepeer/eth/types"
+	"github.com/livepeer/go-livepeer/net"
+
+	"github.com/golang/glog"
+)
+
+type DBOrchestratorPoolCache struct {
+	node *core.LivepeerNode
+}
+
+func NewDBOrchestratorPoolCache(node *core.LivepeerNode) *DBOrchestratorPoolCache {
+	if node.Eth == nil {
+		glog.Error("Could not refresh DB list of orchestrators: LivepeerNode nil")
+		return nil
+	}
+
+	ticker := time.NewTicker(1 * time.Hour)
+	go func(node *core.LivepeerNode) {
+		for {
+			select {
+			case <-ticker.C:
+				orchestrators, err := node.Eth.RegisteredTranscoders()
+				if err != nil {
+					glog.Error("Could not refresh DB list of orchestrators: ", err)
+					return
+				}
+				_, dbOrchErr := CacheDBOrchs(node, orchestrators)
+				if dbOrchErr != nil {
+					glog.Error("Could not refresh DB list of orchestrators: CacheDBOrchs err")
+					return
+				}
+			}
+		}
+	}(node)
+
+	return &DBOrchestratorPoolCache{node: node}
+}
+
+func CacheDBOrchs(node *core.LivepeerNode, orchs []*lpTypes.Transcoder) ([]*common.DBOrch, error) {
+	var dbOrchs []*common.DBOrch
+	if orchs == nil {
+		glog.Error("No new DB orchestrators created: no orchestrators found onchain")
+		return dbOrchs, nil
+	}
+
+	for _, orch := range orchs {
+		dbOrch := EthOrchToDBOrch(orch)
+		if err := node.Database.UpdateOrchs(dbOrch); err != nil {
+			glog.Error("Error updating Orchestrator in DB: ", err)
+			return nil, err
+		}
+		dbOrchs = append(dbOrchs, dbOrch)
+	}
+	return dbOrchs, nil
+
+}
+
+func EthOrchToDBOrch(orch *lpTypes.Transcoder) *common.DBOrch {
+	if orch == nil {
+		return nil
+	}
+	return common.NewDBOrch(orch.ServiceURI, orch.Address.String())
+}
+
+func (dbo *DBOrchestratorPoolCache) GetOrchestrators(numOrchestrators int) ([]*net.OrchestratorInfo, error) {
```

Comment:
Is that a best practice in go? I am fine reorganizing the functions. It does seem like it would improve readability

Reply:
I honestly don't know the idiomatic Go file structure. Found [this thread](https://www.reddit.com/r/golang/comments/2xn1ws/idiomatic_go_ordering_of_functions_in_source_files/) with some thoughts, in which no "official" position was shared.

I very much connect with doing our best to make code read coherently from top to bottom, which is why I like placing public functions before private ones, so that when you reach the private function you already know in what context it's used. I also find it very helpful to make sure all of a type's functions are batched together (they can be in separate files if it makes sense, but they shouldn't be scattered across a file with other functions intermingled). 

Perhaps @j0sh should weigh in here and suggest a standard we should stick to?

Reply:
I don't really have a strong preference about function placement. The proposed approach sounds good.

Personally, I typically start off with keeping function definitions as close as possible to its use to make quick scanning (or even editing) easy, even if this means some interleaving of methods and functions. Of course, there are lots of exceptions to this, and code tends to drift anyway without a conscious effort to keep things organized.

The C programmer in me has always been slightly bemused about defining a function *after* it's used (without a forward declaration) -- it feels a bit like using a variable before it's declared. Lots of languages have this feature though, so why not.